core/lib/zksync_core/src/reorg_detector/mod.rs

```diff
@@ -38,21 +38,31 @@ enum HashMatchError {
         Using an earlier snapshot could help."
     )]
     EarliestL1BatchTruncated(L1BatchNumber),
-    #[error("Internal error")]
-    Internal(#[from] anyhow::Error),
+    #[error("reorg detected, restart the node to revert to the last correct L1 batch #{0}.")]
+    ReorgDetected(L1BatchNumber),
+    #[error("storage error: {0:#}")]
+    Storage(anyhow::Error),
+    #[error("no remote L1 batch")]
+    NoRemoteL1Batch,
```

Comment:
Bikeshedding: While I appreciate that it allows simplifying code quite a bit, IMO, having this error variant exposed publicly makes it more difficult to reason about public `ReorgDetector` methods, because in some cases, this variant isn't really an error. It could make sense to have an internal error enum representing a sum of `Error` and `NoRemoteL1Batch` variants.

Reply:
It is an error in a sense that comparison couldn't be performed. The real issue is that all the methods are using the same error type.